File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/devdb/brand/DevicesAdapter.java
package forpdateam.ru.forpda.ui.fragments.devdb.brand;

import android.graphics.Bitmap;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.assist.FailReason;
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.entity.remote.devdb.Brand;
import forpdateam.ru.forpda.ui.fragments.devdb.DevDbHelper;
import forpdateam.ru.forpda.ui.views.adapters.BaseAdapter;
import forpdateam.ru.forpda.ui.views.adapters.BaseViewHolder;

/**
 * Created by radiationx on 08.08.17.
 */

public class DevicesAdapter extends BaseAdapter<Brand.DeviceItem, DevicesAdapter.DeviceItemHolder> {
    private BaseAdapter.OnItemClickListener<Brand.DeviceItem> itemClickListener;

    public void setItemClickListener(OnItemClickListener<Brand.DeviceItem> itemClickListener) {
        this.itemClickListener = itemClickListener;
    }

    @Override
    public DevicesAdapter.DeviceItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = inflateLayout(parent, R.layout.brand_item);
        return new DevicesAdapter.DeviceItemHolder(v);
    }

    @Override
    public void onBindViewHolder(DevicesAdapter.DeviceItemHolder holder, int position) {
        holder.bind(getItem(position), position);
    }

    class DeviceItemHolder extends BaseViewHolder<Brand.DeviceItem> implements View.OnClickListener, View.OnLongClickListener {
        TextView title;
        TextView rating;
        ImageView image;
        ProgressBar progressBar;

        DeviceItemHolder(View v) {
            super(v);
            title = (TextView) v.findViewById(R.id.item_title);
            rating = (TextView) v.findViewById(R.id.item_rating);
            image = (ImageView) v.findViewById(R.id.item_image);
            progressBar = (ProgressBar) v.findViewById(R.id.progress_bar);
            image.setTag(progressBar);
            rating.setBackground(App.getDrawableAttr(rating.getContext(), R.attr.count_background));
            v.setOnClickListener(this);
            v.setOnLongClickListener(this);
        }

        @Override
        public void bind(Brand.DeviceItem item, int position) {
            title.setText(item.getTitle());
            if (item.getRating() > 0) {
                rating.setText(Integer.toString(item.getRating()));
                rating.getBackground().setColorFilter(DevDbHelper.INSTANCE.getColorFilter(item.getRating()));
                rating.setVisibility(View.VISIBLE);
            } else {
                rating.setVisibility(View.GONE);
            }
            ImageLoader.getInstance().displayImage(item.getImageSrc(), image, new SimpleImageLoadingListener() {
                @Override
                public void onLoadingStarted(String imageUri, View view) {
                    ProgressBar progressBar = (ProgressBar) view.getTag();
                    progressBar.setVisibility(View.VISIBLE);
                }

                @Override
                public void onLoadingCancelled(String imageUri, View view) {
                    ProgressBar progressBar = (ProgressBar) view.getTag();
                    progressBar.setVisibility(View.GONE);
                }

                @Override
                public void onLoadingComplete(String imageUri, View view, Bitmap loadedImage) {
                    ProgressBar progressBar = (ProgressBar) view.getTag();
                    progressBar.setVisibility(View.GONE);
                }

                @Override
                public void onLoadingFailed(String imageUri, View view, FailReason failReason) {
                    ProgressBar progressBar = (ProgressBar) view.getTag();
                    progressBar.setVisibility(View.GONE);
                }
            });
        }

        @Override
        public void onClick(View view) {
            if (itemClickListener != null) {
                itemClickListener.onItemClick(getItem(getLayoutPosition()));
            }
        }

        @Override
        public boolean onLongClick(View view) {
            if (itemClickListener != null) {
                itemClickListener.onItemLongClick(getItem(getLayoutPosition()));
                return true;
            }
            return false;
        }
    }

}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/devdb/brand/DevicesAdapter.kt
package forpdateam.ru.forpda.ui.fragments.devdb.brand

import android.graphics.Bitmap
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import by.kirich1409.viewbindingdelegate.viewBinding
import com.nostra13.universalimageloader.core.ImageLoader
import com.nostra13.universalimageloader.core.assist.FailReason
import com.nostra13.universalimageloader.core.listener.SimpleImageLoadingListener
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.databinding.BrandItemBinding
import forpdateam.ru.forpda.entity.remote.devdb.Brand.DeviceItem
import forpdateam.ru.forpda.extensions.setBackgroundAttr
import forpdateam.ru.forpda.ui.fragments.devdb.DevDbHelper.getColorFilter
import forpdateam.ru.forpda.ui.fragments.devdb.brand.DevicesAdapter.DeviceItemHolder
import forpdateam.ru.forpda.ui.views.adapters.BaseAdapter
import forpdateam.ru.forpda.ui.views.adapters.BaseViewHolder

/**
 * Created by radiationx on 08.08.17.
 */
class DevicesAdapter : BaseAdapter<DeviceItem, DeviceItemHolder>() {

    private var itemClickListener: OnItemClickListener<DeviceItem>? = null

    fun setItemClickListener(itemClickListener: OnItemClickListener<DeviceItem>?) {
        this.itemClickListener = itemClickListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeviceItemHolder {
        val v = inflateLayout(parent, R.layout.brand_item)
        return DeviceItemHolder(v)
    }

    override fun onBindViewHolder(holder: DeviceItemHolder, position: Int) {
        holder.bind(getItem(position), position)
    }

    inner class DeviceItemHolder(v: View) : BaseViewHolder<DeviceItem>(v),
        View.OnClickListener, OnLongClickListener {

        private val binding by viewBinding<BrandItemBinding>()

        init {
            binding.itemRating.setBackgroundAttr(R.attr.count_background)
            v.setOnClickListener(this)
            v.setOnLongClickListener(this)
        }

        override fun bind(item: DeviceItem, position: Int) {
            binding.itemTitle.text = item.title
            if (item.rating > 0) {
                binding.itemRating.text = item.rating.toString()
                binding.itemRating.background.colorFilter = getColorFilter(item.rating)
                binding.itemRating.visibility = View.VISIBLE
            } else {
                binding.itemRating.visibility = View.GONE
            }
            ImageLoader.getInstance()
                .displayImage(item.imageSrc, binding.itemImage, object : SimpleImageLoadingListener() {
                    override fun onLoadingStarted(imageUri: String, view: View) {
                        binding.progressBar.visibility = View.VISIBLE
                    }

                    override fun onLoadingCancelled(imageUri: String, view: View) {
                        binding.progressBar.visibility = View.GONE
                    }

                    override fun onLoadingComplete(
                        imageUri: String,
                        view: View,
                        loadedImage: Bitmap
                    ) {
                        binding.progressBar.visibility = View.GONE
                    }

                    override fun onLoadingFailed(
                        imageUri: String,
                        view: View,
                        failReason: FailReason
                    ) {
                        binding.progressBar.visibility = View.GONE
                    }
                })
        }

        override fun onClick(view: View) {
            if (itemClickListener != null) {
                itemClickListener!!.onItemClick(getItem(layoutPosition))
            }
        }

        override fun onLongClick(view: View): Boolean {
            if (itemClickListener != null) {
                itemClickListener!!.onItemLongClick(getItem(layoutPosition))
                return true
            }
            return false
        }
    }
}
